### DISCUSSION TITLE
QOL Idea for Consideration

### DISCUSSION THREAD

**Discussion by ScissorShiv** [diasurgical/DevilutionX#5937] [Ideas]
First off I'd like to say how amazing this project is for a game that is not only a significant landmark in gaming, but also an amazing game to play and experience to this day.  Cheers!

When playing this game in the past, I obviously was restricted to using a keyboard and mouse.  The addition of controller support is absolutely a welcomed sight.  The QOL addition I'd like to suggest is something that I was able to do with the mouse and keyboard, but not the controller.  This was the ability to quickly swap weapons during game play (even during combat when openings arouse).

What I'd like to suggest is the addition of an ability given to all characters that allows for being able to swap weapons.  I very much do not want to change the Diablo 1 game play experience and offer the following manner in which to implement the idea with, what I feel is no downside, as follows:

1. The weapon swap would be an ability that each of the characters has.
2. The ability would be handled just as other character abilities, found in the spell HUD as a gold icon, and able to be queued in the speed book hotkeys.
3. When the ability is selected, the cast button would be used to initiate the ability (just as other abilities).
4. The ability would allow the character to swap into a different weapon preset, that has been selected beforehand in the equipment menu.
5. The ability would have a "cast" duration just as other spells/ abilities, that could be cancelled and interrupted by receiving stunning damage (just like spells work currently).
6. The alternate weapons need to be in the players inventory for the swap to occur.
7. The weapons being swapped out would automatically be placed in the characters open inventory space, if there is not enough room in the characters inventory for one or both of the items, it or they are dropped to the ground.  
8. The left and right hand would be handled separately, so that if nothing is preset in one hand it would simply keep the current item equipped (ie swapping your sword for a mace while retaining your shield in the right hand) or place it into the characters inventory if needed (ie swapping from a sword and shield to a two handed sword)

Other thoughts:

The time it takes to swap weapons could be tweaked to be slightly longer for certain characters than other (just like how it takes longer to cast a spell for a warrior than that of the sorcerer) - not necessary, but could be a unique way to add some difference in the classes.

A toggle in the options menu to enable or disable this modification.

Final comments:

I truly feel that this addition would not alter the Diablo experience at all.  This is something that could already be done, but it is discouraged by being clunky and needing to open panels, drag and dropping...etc, detracting from the immersion of the game.  By creating this less intrusive manner in accomplishing the same end, the player is now encouraged to do so allowing for more strategy when facing down a level.  The idea of seeing a horde of undead skulking toward you and being able to swap out your sword for a mace to take advantage of the damage increase without needing to open panels and rearrange items in your inventory just sounds amazing.  I understand that something like this was implemented in Diablo 2 and some other mods, but the reason I detailed the manner I recommend above is to do everything possible to not alter the Diablo 1 experience.  The swap requires the use of a skill that takes a brief moment to execute and it does not create additional inventory room, meaning you have to weigh out whether it is worth carrying the extra gear, whether you want to take up one of your hotkeys for the ability...etc.   At any rate, I can't thank those who have created this mod enough and appreciate your time and consideration of this addition to an already amazing effort.

Cheers!

**Comment by yuripourre** [comment]
@ScissorShiv I think this idea makes sense. It's something that players do (swapping weapons) so it's not an entire new mechanic.

Having said that, could you elaborate how do you envision this "preset" mechanism? The player needs to select one or two items to be the ones to be switched. How do you envision that?

This by itself adds a lot of complexity to the QoL.

**Reply by ScissorShiv** [reply]
So the way I see it in my own mind is that the left and right hand represented in the inventory panel would have two tabs.  We'll call them tab A and tab B just for this example.  

So lets say your in a fresh game as the warrior and you have your shield, short sword, and club.  As normal, you would equip your short sword in your right hand in tab A, then your shield in the left hand also in tab A.  From here you would click tab B on the right hand and place your club there (the club however remains in the players inventory, or is at least represented so as to not change the amount of available inventory).  Now as your navigating the first level of the cathedral, you decide to swap from your short sword to your club.  You cue up your weapon swap ability, activate it with the cast button, the ability animation with sound effect occurs, and the swap weapon skill toggles from whats in tab A to what is in tab B.  The weapon in tab A (short sword) is now either placed or represented in the characters inventory.  The shield in the left hand remains since there was nothing cued in tab B for that hand.  If the weapon swap ability is activated again, then the same process occurs toggling back to the tab A weapons. 

This is just what I am envisioning in my head based on my understanding of D1 and this concept iterated in other games.  There may well be a better way to do this, the premise here being an attempt to duplicate weapon swapping in D1 in a friendlier and more effective fashion.

**Reply by julealgon** [reply]
What happens if the weapon you are equipped with is larger than the one currently in the inventory that is the target for the swap, and your currently equipped weapon doesn't fit the inventory?

**Reply by AJenbo** [reply]
I guess the hero just complains about lack of space instead of swapping.

**Reply by ScissorShiv** [reply]
In the event that this occurs, then the offending equipment (ie the equipment that does not fit in the inventory) would be dropped to the ground.  This instance would occur most often when swapping from a sword/mace and shield (9 inventory spaces) into a two handed weapon such as a bow/staff/axe/two-handed sword/maul (6 inventory spaces), and the rest of the inventory space is filled or unorganized.  There is the possibility of this occurring when swapping from a sword (3 inventory spaces) to a dagger (2 inventory spaces).  This is what players would often do when switching weapons and this issued occurred.  Using the same example, the player is swapping from a sword/shield into a two-handed weapon.  The inventory panel is opened, the two-handed weapon is dragged and dropped into the right hand, the sword is automatically placed into the inventory while the shield is affixed to the cursor.  The player then clicks off screen and drops the shield to the ground.

**Reply by AJenbo** [reply]
Can't wait for the bug reports "Important gear vanishing with no trace randomly", then weeks of debugging and not realizing they didn't notice when the feature dropped there good gear instead of the junk they picked up during that session.

**Reply by ScissorShiv** [reply]
I suppose the alternative would be exactly what you mentioned, the swap doesn't occur, and the character says "now where would I put this".

**Comment by AJenbo** [comment]
I think this is the best proposal for how to handle this, but I'm having a hard time understanding how the player assigns pairs of items to the ability?

It also very easily slips in to mod territory.

**Reply by StephenCWills** [reply]
I would just have the player equip the items. Those items could be automatically assigned to whatever is the currently active weapon set. This also has the advantage of allowing you to use the same shield for the two different weapon sets, for example. It makes the logic in the code pretty complicated, but still pretty intuitive for the player.

**Reply by AJenbo** [reply]
So equip your gear, cast the remember-spell to remember, then switch gear and cast again to remember another set?
Later cast the swap-spell to swap between the two last remembered items?

One issue here is that the items aren't tracked by id but by location, so we would probably need to search for the item. So we need the spell to store both seed and idx for each item to be able to find them.

Does it just remember the two last setes and push out the oldest one?

**Reply by ScissorShiv** [reply]
Wouldn't it be easier to create two instances of the left and right hand,  one active and one dormant.  Then the swap weapon skill acts like a switch to toggle between the two instances, this way the location would still be correct.

**Reply by AJenbo** [reply]
So maintain the first instance as is, but have the second one be a "ghost" instance that only references an item from the inventory?

That wouldn't allow you to use the same shield i think.

**Reply by ScissorShiv** [reply]
If the left and right hand were handled separately, then in the event that there is no item assigned to one of the hands, then the item in that hand would persist.  This shouldn't cause an issue with two-handed weapons, as they create their likeness in both hands seen as a translucent version of itself with a red background in the off-hand.

**Reply by ScissorShiv** [reply]
I don't know if this would make things easier, but I think it may be better to reverse the "ghost" instance so that the items are in the hand slots and the "ghost" instances are in the inventory to represent the space that is being used for the items not active.

**Reply by StephenCWills** [reply]
> So equip your gear, cast the remember-spell to remember, then switch gear and cast again to remember another set?
> Later cast the swap-spell to swap between the two last remembered items?

I wasn't thinking that there'd be a separate "Remember" spell. I figured there'd just be the one "Swap" action/spell. Whatever you currently have equipped is the set of equipment for the active weapon set. When you swap weapon sets, it would change your equipment around to whatever is recorded in the second weapon set. Now that you have swapped weapon sets, you can change your equipment around to edit that weapon set. So there would be no extra step besides using the "Swap" action/spell and changing your equipment.

> One issue here is that the items aren't tracked by id but by location, so we would probably need to search for the item. So we need the spell to store both seed and idx for each item to be able to find them.

I agree with that assessment, but since the solutions we prefer will require that the inactive weapon set should be located in the inventory, I'm not sure how you'd get around that.

One other issue here is what to do if you remove an item from your inventory that is currently assigned to the inactive weapon set. If you want to be able to use the same shield in two weapon sets, you'd have to give the player some leeway in terms of moving the item to the cursor at least without deleting it from the weapon set. Probably the simplest thing to do in the game logic would be to ignore when the item is removed from the inventory and just search for it every time you do a weapon swap.

**Reply by StephenCWills** [reply]
> Wouldn't it be easier to create two instances of the left and right hand, one active and one dormant. Then the swap weapon skill acts like a switch to toggle between the two instances, this way the location would still be correct.

No, I think it would not be easier at all. For that, you'd have to do something tricky like maintaining two copies of the item so it can exist in the dormant hand and the inventory simultaneously. And you'd still have to search the inventory for it when you swap weapon sets either way.

**Reply by StephenCWills** [reply]
> I don't know if this would make things easier, but I think it may be better to reverse the "ghost" instance so that the items are in the hand slots and the "ghost" instances are in the inventory to represent the space that is being used for the items not active.

And then, if the player needed to move the ghost around to make room in their inventory, I suppose the cursor item should be made into a ghost cursor? Seems quite a bit harder, actually.

Generally, it's better not to overcomplicate, conceptually. When the items are not in the active weapon set, they are actually in the inventory, and that allows us to interact with them in the same way as any other item in the inventory. They are not in some ephemeral space that takes up space in the inventory which should therefore be made to have special cases for various player interactions.

**Reply by AJenbo** [reply]
> in the event that there is no item assigned to one of the hands, then the item in that hand would persist

What if i have it so that i can switch to a sword but no shield for dealing with ranged enemies so that i don't get stunned from blocking?

**Reply by ScissorShiv** [reply]
Ok, so in order to keep this addition in the realm of QoL and out of the Mod arena, two points come to mind:

1. The swap needs to not be instantaneous as to reflect the short but meaningful time it would take to switch one's weapons.  The concept of using a ability with cast rate...etc seems to adequately address this point.
2. The items being swapped in, need to be present in the characters inventory as to not modify the inventory carry capacity limitations.  This appears to be the tricky part of the endeavor.

What if you were to make it so that the swap weapon ability would swap the currently equipped item with the item specifically place in the inventory located in the 2x3 section to the immediate left for the right hand.  To the immediate right of that section would be the 2x3 area of the inventory designated for the left hand.  So what I'm suggesting is that an area within the inventory would serve as a ready area for the left and right hands, and the player would need to stage their weapon swap choices within these designated sections.  Upon activating the weapon swap, the game would need to ensure that the items in these areas can legally be equipped.  If they cannot, then the character would respond with "I can't do that right now".  It would be up to the player to correctly stage their items in their inventory so that the swap would be legal and not fail when used.  This does bring up several troubleshooting questions; however it may be a step towards complying with point #2 above.   

Things to troubleshoot:

1. How to handle situations where potions or scroll are placed within these designated areas. (could the game ignore all items not able to be equipped to the hand?  So the swap ability searches the designated area and if any legal items are available it would swap them and ignore the non-legal items)
2. What if the player has a sword in the right hand staging area and is swapping out a 2-handed maul, but the staging area also has a scroll in it - I would assume that the swap would not occur.  This does make it a bit difficult for the player as they would have to be vigilant about maintaining a swap conducive inventory, 
3. How to retain a shield while swapping a mace for a sword or vice versus.  (I still like the idea that if the staging area within the inventory has no legal replacement then the item in hand persists)
4. As Ajenbo mentioned, what if you want to retain your sword, but stow your shield to avoid being block-stunned.  (This may be a limitation with the ability, but could be compensated for by swapping to a 2-handed sword which would un-equip the shield - not ideal but it is an action a player could make to work around the limitation.)   

I realize that this may or may not be the best route to go, but it's worth kicking around to see if there may be something to it or lead to a better/ more refined manner to do it.  The list of troubleshoot items I mentioned is not comprehensive, just a quick few that sprang to mind at the time of writing.

**Reply by AJenbo** [reply]
not very intuitive, but comparatively easy to implement and avoids going many of the things that makes it feel more like a mod

**Reply by StephenCWills** [reply]
I don't know that it's comparatively easy to implement. What if you have both a Short Sword and a Dagger in the staging area for the left hand?

**Reply by AJenbo** [reply]
i would pick the first valid that i find in the area, not very intuitive, so not saying it's all good :)

**Reply by StephenCWills** [reply]
Well, personally I think the hardest part of any suggestion for this feature is computing the space available for the items you need to dequip and then proceeding to swap all the items around. I guess it does make things easier if you only have to look at the 2x3 space below each hand. I think it does make things significantly more complicated for the player, though, if we don't take special care to avoid those areas when auto-populating the inventory.

All this discussion has got me wanting to try my hand at implementing something, but I feel I need to focus my energy on gamepad for the upcoming release.

**Reply by AJenbo** [reply]
Yeah a good place to start would be to implement in-place swapping for shift+click. The reason we don't have that atm is also because of the difficulty in calculating the available space ahead of the move.
